pkg/ccl/streamingccl/streamingest/replication_execution_details.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/cockroachdb/cockroach/pkg/util/span"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/redact"
+)
+
+type frontierExecutionDetails struct {
+	srcInstanceID  base.SQLInstanceID
+	destInstanceID base.SQLInstanceID
+	span           string
+	frontierTS     string
+	behindBy       redact.SafeString
+}
+
+// constructSpanFrontierExecutionDetails constructs the frontierExecutionDetails
+// using the initial partition specs that map spans to the src and dest
+// instances, and a snapshot of the current state of the frontier.
+//
+// The shape of the spans tracked by the frontier can be different from the
+// initial partitioned set of spans. To account for this, for each span in the
+// initial partition set we want to output all the intersecting sub-spans in the
+// frontier along with their timestamps.
+func constructSpanFrontierExecutionDetails(
+	partitionSpecs execinfrapb.StreamIngestionPartitionSpecs,
+	frontierSpans execinfrapb.FrontierEntries,
+) ([]frontierExecutionDetails, error) {
+	f, err := span.MakeFrontier()
+	if err != nil {
+		return nil, err
+	}
+	for _, rs := range frontierSpans.ResolvedSpans {
+		if err := f.AddSpansAt(rs.Timestamp, rs.Span); err != nil {
+			return nil, err
+		}
+	}
+
+	now := timeutil.Now()
+	res := make([]frontierExecutionDetails, 0)
+	for _, spec := range partitionSpecs.Specs {
+		for _, sp := range spec.Spans {
+			f.SpanEntries(sp, func(r roachpb.Span, timestamp hlc.Timestamp) (done span.OpResult) {
+				res = append(res, frontierExecutionDetails{
+					srcInstanceID:  spec.SrcInstanceID,
+					destInstanceID: spec.DestInstanceID,
+					span:           r.String(),
+					frontierTS:     timestamp.GoTime().String(),
+					behindBy:       humanizeutil.Duration(now.Sub(timestamp.GoTime())),
+				})
+				return span.ContinueMatch
+			})
+		}
+
+		// Sort res on the basis of srcInstanceID, destInstanceID.
+		sort.Slice(res, func(i, j int) bool {
+			if res[i].srcInstanceID != res[j].srcInstanceID {
+				return res[i].srcInstanceID < res[j].srcInstanceID
+			}
+			if res[i].destInstanceID != res[j].destInstanceID {
+				return res[i].destInstanceID < res[j].destInstanceID
+			}
+			return res[i].span < res[j].span
+		})
+	}
+
+	return res, nil
+}
+
+// generateSpanFrontierExecutionDetailFile generates and writes a file to the
+// job_info table that captures the mapping from:
+//
+// # Src Instance	| Dest Instance	| Span | Frontier Timestamp	| Behind By
+//
+// This information is computed from information persisted by the
+// stream ingestion resumer and frontier processor. Namely:
+//
+// - The StreamIngestionPartitionSpec of each partition providing a mapping from
+// span to src and dest SQLInstanceID.
+// - The snapshot of the frontier tracking how far each span has been replicated
+// up to.
+func generateSpanFrontierExecutionDetailFile(
+	ctx context.Context, execCfg *sql.ExecutorConfig, ingestionJobID jobspb.JobID, skipBehindBy bool,
+) error {
+	var sb bytes.Buffer
```

Comment:
given the possibility of txn retries, i think `sb` and `w` should be created within the txn, right?

Reply:
you are very right, nice catch